include/xtensor-blas/xlinalg.hpp

```diff
@@ -1514,6 +1514,172 @@ namespace linalg
+            }
+            auto a_mat = result_type::from_shape({keep_a_len, sum_len});
+            std::copy(a.storage().begin(), a.storage().end(), a_mat.storage().begin());
+            auto b_mat = result_type::from_shape({sum_len, keep_b_len});
+            std::copy(b.storage().begin(), b.storage().end(), b_mat.storage().begin());
+
+            result = dot(a_mat, b_mat);
+            result.reshape(result_shape);
+        }
+        return result;
+    }
+
+    /**
+     * @brief Compute tensor dot product along specified axes for arrays
+     *
+     * Compute the sum of products along the axes \em ax_a for a and \em ax_b for b
+     *
+     * @param xa input array
+     * @param xb input array
+     * @param ax_a axes to sum over for \em a
+     * @param ax_b axes to sum over for \em b
+     * @return resulting array
+     */
+    template <class T, class O>
+    auto tensordot(const xexpression<T>& xa, const xexpression<O>& xb, const std::vector<std::size_t>& ax_a,
+                   const std::vector<std::size_t>& ax_b)
+    {
+        const auto& a = xa.derived_cast();
+        const auto& b = xb.derived_cast();
+        XTENSOR_ASSERT(ax_a.size() == ax_b.size());
+        XTENSOR_ASSERT(ax_a.size() < a.dimension());
+        XTENSOR_ASSERT(ax_b.size() < b.dimension());
+        std::size_t n_ax = ax_a.size();
+        for (std::size_t i = 0; i < n_ax; ++i)
+        {
+            XTENSOR_ASSERT(ax_a[i] < a.dimension());
+            XTENSOR_ASSERT(ax_b[i] < b.dimension());
+        }
+
+        // Move the axes to sum over to the end of a
+        std::vector<std::size_t> newaxes_a;
+        std::vector<std::size_t> result_shape;
+        auto newax_a_inserter = std::back_inserter(newaxes_a);
+        for (std::size_t i = 0; i < a.dimension(); ++i)
+        {
+            auto a_ax_it = std::find(ax_a.begin(), ax_a.end(), i);
+            // first pass if i is not in ax_a, add to the newaxes_a, and the result shape
+            if (a_ax_it == ax_a.end())
+            {
+                newax_a_inserter = i;
+            }
+        }
+        for (auto a_ax_it = ax_a.begin(); a_ax_it != ax_a.end(); ++a_ax_it)
+        {
+          newax_a_inserter = *a_ax_it;
+
+        }
+
+        // Move the axes to sum over to the start of b
+        std::vector<std::size_t> newaxes_b;
+        auto newax_b_inserter = std::back_inserter(newaxes_b);
+        for (auto b_ax_it = ax_b.begin(); b_ax_it != ax_b.end(); ++b_ax_it)
+        {
+          newax_b_inserter = *b_ax_it;
+        }
+        for (std::size_t i = 0; i < b.dimension(); ++i)
+        {
+            auto b_ax_it = std::find(ax_b.begin(), ax_b.end(), i);
+            // seccond pass if i is not in ax_b add to the newaxes_b, and the result shape
+            if (b_ax_it == ax_b.end())
+            {
+                newax_b_inserter = i;
+            }
+        }
+        // not 100% sure why, but If I don't eval this I get the wrong result.
```

Comment:
That sounds like a problem in ... `dot` maybe? I will have to look into it but for now we can keep the `eval` I guess.